README.txt
School research project
------------------------
I created this package for my school research project.
I want to create an AI that can predict if a human where to get a heart failure.
For this I created a very long python file but I wanted to make this easier to understand.
So, I created this package to get a shorter file and make it easier to understand.

How to use
--------------
This package uses the scikit-learn package and therefore some classes like the MLPRegressor are the same.

1. Import the package: "import zeroone_ai"
2. Choose wich class you are going to use: "from zeroone_ai import MLPRegressor
3. (More comming in the future)

License
--------------
- scikit-learn (new BSD) https://pypi.org/project/scikit-learn/
- pandas (BSD 3-Clause License) https://github.com/pandas-dev/pandas/blob/master/LICENSE
- matplotlib (Python Software Foundation License (PSF)) https://pypi.org/project/matplotlib/3.4.3/
- numpy (BSD License (BSD)) https://github.com/numpy/numpy/blob/main/LICENSE.txt

CHANGELOG.txt
Change Log
===========

0.0.11 (18-11-2021)
-------------------
- You can now predict unknown data when using the new mapping feature

0.0.10 (27-10-2021)
-------------------
- The LogisticRegressor bug has been fixed 
- Plots will now always show

0.0.9 (27-10-2021)
-------------------
- The AI now predicts a zero or an one
- Added MLPClassifier and LogisticRegressor

0.0.8 (21-10-2021)
-------------------
- Added epoch training

0.0.7 (20-10-2021)
-------------------
- Fixed a bug where the library couldn't be imported

0.0.4 (19-10-2021)
-------------------
- Fixed some bugs and provided a better README file

0.0.3 (19-10-2021)
-------------------
- Fixed some bugs and provided a better README file

0.0.2 (19-10-2021)
-------------------
- Test if updating works... It does

0.0.1 (19-10-2021)
-------------------
- First Release

src/zerooneai/test.py
from zerooneai import MLPRegressor

dataset = 'cardio_train.csv'
input_data = ['age','gender','height','weight','ap_hi','ap_lo','cholesterol','gluc','smoke','alco','active']
label = 'cardio'
model = MLPRegressor(dataset,input_data,label)

hidden_layer_sizes = [20,20,10]
model.train(hidden_layer_sizes)

print(model.accuracy())

length = 20
model.plot(length)

src/zerooneai/zerooneai.py
import numpy as np
from numpy.core.numeric import tensordot
import pandas as pd
import pickle
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor as MLP
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import plot_confusion_matrix

class MLPRegressor:
    def __init__(self,dataset,input_data,label,test_size=0.2,random_state=42):
        self.label = label
        dataset = pd.read_csv(dataset)
        input_data = dataset[input_data]
        input_data = (input_data - input_data.mean()) / (input_data.max() - input_data.min()) # stap 3
        label = dataset[self.label]

        self.input_data_train, self.input_data_test, self.labels_train, self.labels_test = train_test_split(input_data,label,test_size=test_size,random_state=random_state)

    def accuracy(self):
        lengte = len(self.true_labels)
        all_errors = 0
        for pred_label,true_label in zip(self.pred_labels,self.true_labels):
            if pred_label < 0.5:
                pred_label = 0
            else:
                pred_label = 1
            if pred_label == true_label:
                all_errors += 1
            

        MAE = all_errors / lengte
        return MAE

    def plot(self,length,yrange=None,ax=None):
        if ax == None:
            fig, ax = plt.subplots()
        index = np.arange(length)
        bar_width = 0.25
        opacity = 0.8

        rects1 = ax.bar(index, self.true_labels[:length], bar_width, alpha=opacity,color='r',label='True_labels')
        rects2 = ax.bar(index + bar_width, self.pred_labels[:length], bar_width,alpha=opacity,color='b',label='Pred_labels')
        rects2 = ax.bar(index + 2*bar_width, abs(self.true_labels[:length]-self.pred_labels[:length]), bar_width,alpha=opacity,color='g',label='Absolute Error')
        ax.ticklabel_format(style='plain')
        ax.set_xlabel('')
        ax.set_ylabel('')
        ax.set_title(self.label)
        if yrange != None:
            ax.set_ylim(yrange)
        ax.legend()

        if ax == None:
            plt.show()

    def train(self,hidden_layer_sizes):
        self.model = MLP(hidden_layer_sizes)
        self.model.fit(self.input_data_train,self.labels_train)
        
        self.pred_labels = self.model.predict(self.input_data_test)
        self.true_labels = self.labels_test


zeroone_ai/__init__.py
import numpy as np
from numpy.core.numeric import tensordot
import pandas as pd
import pickle
import matplotlib.pyplot as matplot
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor as MLPR
from sklearn.linear_model import LogisticRegression as LR
from sklearn.neural_network import MLPClassifier as MLPC
from sklearn.metrics import plot_confusion_matrix

def accuracy(pred_labels,true_labels):
        lengte = len(true_labels)
        all_errors = 0
        for pred_label,true_label in zip(pred_labels,true_labels):
            if pred_label == true_label:
                all_errors += 1
            

        MAE = all_errors / lengte
        return MAE

def normalize(df, mapping):
    result = df.copy()
    for feature_name in df.columns:
        max_value = mapping[feature_name]['max']
        min_value = mapping[feature_name]['min']
        result[feature_name] = (df[feature_name] - min_value) / (max_value - min_value)
    return result


def plt(true_labels,pred_labels,label,length,yrange,ax):
    if ax == None:
        fig, ax = matplot.subplots()
    index = np.arange(length)
    bar_width = 0.25
    opacity = 0.8

    rects1 = ax.bar(index, true_labels[:length], bar_width, alpha=opacity,color='r',label='True_labels')
    rects2 = ax.bar(index + bar_width, pred_labels[:length], bar_width,alpha=opacity,color='b',label='Pred_labels')
    rects2 = ax.bar(index + 2*bar_width, abs(true_labels[:length]-pred_labels[:length]), bar_width,alpha=opacity,color='g',label='Absolute Error')
    ax.ticklabel_format(style='plain')
    ax.set_xlabel('')
    ax.set_ylabel('')
    ax.set_title(label)
    if yrange != None:
        ax.set_ylim(yrange)
    ax.legend()

    if ax == None:
        matplot.show()

    matplot.show()

class MLPRegressor:
    def __init__(self,dataset,input_data,label,test_size=0.2,random_state=42,mapping=None):
        self.label = label
        self.mapping = mapping
        dataset = pd.read_csv(dataset)
        input_data = dataset[input_data]
        if mapping == None:
            input_data = (input_data - input_data.mean()) / (input_data.max() - input_data.min())
            self.mean = input_data.mean()
            self.max = input_data.max()
            self.min = input_data.min()
        else:
            input_data = normalize(input_data, mapping)
        label = dataset[self.label]

        self.data_train, self.data_test, self.labels_train, self.labels_test = train_test_split(input_data,label,test_size=test_size,random_state=random_state)

    def plot(self,length,yrange=None,ax=None):
        pred_labels = self.model.predict(self.data_test)
        true_labels = self.labels_test
        pred_labels = np.round(np.clip(pred_labels,0,1))

        plt(true_labels,pred_labels,self.label,length,yrange,ax)

    def prediction(self, prediction_data):
        if self.mapping == None:
            prediction_data = (prediction_data - self.mean) / (self.max - self.min)
        else:
            prediction_data = normalize(prediction_data, self.mapping)
        pred_label = self.model.predict(prediction_data)
        return pred_label

    def train(self,hidden_layer_sizes):
        self.model = MLPR(hidden_layer_sizes)
        self.model.fit(self.data_train,self.labels_train)
        
        pred_labels = self.model.predict(self.data_test)
        true_labels = self.labels_test
        pred_labels = np.round(np.clip(pred_labels,0,1))

        self.accuracy = accuracy(pred_labels,true_labels)
        print('Het model heeft een nauwkeurigheid van {}.'.format(self.accuracy))

    def epochtrain(self,hidden_layer_sizes,epochs,num_data):
        self.model = MLPR(hidden_layer_sizes,max_iter=1,warm_start=True)

        train_accs = []
        test_accs = []

        for epoch in range(epochs):
            self.model.fit(self.data_train[:num_data], self.labels_test[:num_data])

            pred_labels = self.model.predict(self.data_train[:num_data])
            true_labels = self.labels_train[:num_data]
            train_acc = accuracy(pred_labels,true_labels)
            train_accs.append(train_acc)

            pred_labels = self.model.predict(self.data_test[:1000])
            true_labels = self.labels_test[:1000]
            test_acc = accuracy(pred_labels,true_labels)
            test_accs.append(test_acc)

        matplot.plot(train_accs,label='Train acc')
        matplot.plot(test_accs,label='Test acc')
        matplot.xlabel('Epoch')
        matplot.ylabel('Accuracy')
        matplot.ylim(0.1)
        matplot.legend()
        matplot.plot()
        matplot.show()

class MLPClassifier:
    def __init__(self,dataset,input_data,label,test_size=0.2,random_state=42,mapping=None):
        self.label = label
        self.mapping = mapping
        dataset = pd.read_csv(dataset)
        input_data = dataset[input_data]
        if mapping == None:
            input_data = (input_data - input_data.mean()) / (input_data.max() - input_data.min())
            self.mean = input_data.mean()
            self.max = input_data.max()
            self.min = input_data.min()
        else:
            input_data = normalize(input_data, mapping)
        label = dataset[self.label]

        self.data_train, self.data_test, self.labels_train, self.labels_test = train_test_split(input_data,label,test_size=test_size,random_state=random_state)

    def plot(self,length,yrange=None,ax=None):
        pred_labels = self.model.predict(self.data_test)
        true_labels = self.labels_test
        pred_labels = np.round(np.clip(pred_labels,0,1))

        plt(true_labels,pred_labels,self.label,length,yrange,ax)
    
    def prediction(self, prediction_data):
        if self.mapping == None:
            prediction_data = (prediction_data - self.mean) / (self.max - self.min)
        else:
            prediction_data = normalize(prediction_data, self.mapping)
        pred_label = self.model.predict(prediction_data)
        return pred_label

    def train(self,hidden_layer_sizes):
        self.model = MLPC(hidden_layer_sizes)
        self.model.fit(self.data_train,self.labels_train)
        
        pred_labels = self.model.predict(self.data_test)
        true_labels = self.labels_test
        pred_labels = np.round(np.clip(pred_labels,0,1))

        self.accuracy = accuracy(pred_labels,true_labels)
        print('Het model heeft een nauwkeurigheid van {}.'.format(self.accuracy))

    def epochtrain(self,hidden_layer_sizes,epochs,num_data):
        self.model = MLPC(hidden_layer_sizes,max_iter=1,warm_start=True)

        train_accs = []
        test_accs = []

        for epoch in range(epochs):
            self.model.fit(self.data_train[:num_data], self.labels_test[:num_data])

            pred_labels = self.model.predict(self.data_train[:num_data])
            true_labels = self.labels_train[:num_data]
            train_acc = accuracy(pred_labels,true_labels)
            train_accs.append(train_acc)

            pred_labels = self.model.predict(self.data_test[:1000])
            true_labels = self.labels_test[:1000]
            test_acc = accuracy(pred_labels,true_labels)
            test_accs.append(test_acc)

        matplot.plot(train_accs,label='Train acc')
        matplot.plot(test_accs,label='Test acc')
        matplot.xlabel('Epoch')
        matplot.ylabel('Accuracy')
        matplot.ylim(0.1)
        matplot.legend()
        matplot.plot()
        matplot.show()

class LogisticRegressor:
    def __init__(self,dataset,input_data,label,test_size=0.2,random_state=42,mapping=None):
        self.label = label
        self.mapping = mapping
        dataset = pd.read_csv(dataset)
        input_data = dataset[input_data]
        if mapping == None:
            input_data = (input_data - input_data.mean()) / (input_data.max() - input_data.min())
            self.mean = input_data.mean()
            self.max = input_data.max()
            self.min = input_data.min()
        else:
            input_data = normalize(input_data, mapping)
        label = dataset[self.label]

        self.data_train, self.data_test, self.labels_train, self.labels_test = train_test_split(input_data,label,test_size=test_size,random_state=random_state)

    def plot(self,length,yrange=None,ax=None):
        pred_labels = self.model.predict(self.data_test)
        true_labels = self.labels_test
        pred_labels = np.round(np.clip(pred_labels,0,1))

        plt(true_labels,pred_labels,self.label,length,yrange,ax)

    def prediction(self, prediction_data):
        if self.mapping == None:
            prediction_data = (prediction_data - self.mean) / (self.max - self.min)
        else:
            prediction_data = normalize(prediction_data, self.mapping)
        pred_label = self.model.predict(prediction_data)
        return pred_label

    def train(self):
        self.model = LR()
        self.model.fit(self.data_train,self.labels_train)
        
        pred_labels = self.model.predict(self.data_test)
        true_labels = self.labels_test
        pred_labels = np.round(np.clip(pred_labels,0,1))

        self.accuracy = accuracy(pred_labels,true_labels)
        print('Het model heeft een nauwkeurigheid van {}.'.format(self.accuracy))

